### PR TITLE
JsonDocument should allow the UTF-8 content-BOM from Stream inputs

### PR DESCRIPTION
**Pull request by bartonjs**
Based on feedback from ASP.Net moving to JsonDocument (https://github.com/aspnet/Extensions/pull/1028#discussion_r257084498)

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 1]
```diff
@@ -12,6 +12,8 @@ namespace System.Text.Json
 {
     public sealed partial class JsonDocument
     {
+        private static readonly byte[] Utf8Bom = { 0xEF, 0xBB, 0xBF };
```
Comment (benaadams) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]:
```suggestion
        private static ReadOnlySpan<byte> Utf8Bom => new byte[] { 0xEF, 0xBB, 0xBF };
```
Its what all the cool kids are doing these days: https://vcsjones.com/2019/02/01/csharp-readonly-span-bytes-static/

Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: I was going to say that my concern with that would be that it might impact the source-package; but it looks like other things in this library already do it: https://github.com/dotnet/corefx/blob/d856beaad11581ff9c6100e73ceb427eb1f1febf/src/System.Text.Json/src/System/Text/Json/JsonConstants.cs#L26-L28

So... reasonable suggestion :smile:

Comment (benaadams) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]:
```suggestion
        private static ReadOnlySpan<byte> Utf8Bom => new byte[] { 0xEF, 0xBB, 0xBF };
```
Its what all the cool kids are doing these days: https://vcsjones.com/2019/02/01/csharp-readonly-span-bytes-static/

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 2]
```diff
@@ -308,6 +310,27 @@ private static ArraySegment<byte> ReadToEnd(Stream stream)
...
+                        rented,
+                        written,
+                        Utf8Bom.Length - written);
+
+                    written += lastRead;
+                } while (lastRead > 0 && written < Utf8Bom.Length);
+
+                // If we have 3 bytes, and they're the BOM, reset the write position to 0.
+                if (written == Utf8Bom.Length &&
+                    Utf8Bom.AsSpan().SequenceEqual(rented.AsSpan(0, Utf8Bom.Length)))
```
Comment (benaadams) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: With above
```suggestion
                    Utf8Bom.SequenceEqual(rented.AsSpan(0, Utf8Bom.Length)))
```

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 3]
```diff
@@ -367,6 +390,28 @@ private static async
...
+                        written,
+                        Utf8Bom.Length - written,
+                        cancellationToken).ConfigureAwait(false);
+
+                    written += lastRead;
+                } while (lastRead > 0 && written < Utf8Bom.Length);
+
+                // If we have 3 bytes, and they're the BOM, reset the write position to 0.
+                if (written == Utf8Bom.Length &&
+                    Utf8Bom.AsSpan().SequenceEqual(rented.AsSpan(0, Utf8Bom.Length)))
```
Comment (benaadams) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]:
```suggestion
                    Utf8Bom.SequenceEqual(rented.AsSpan(0, Utf8Bom.Length)))
```
With above

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: FWIW: This change (and some experimenting with extra locals) makes the synchronous ReadToEnd slightly longer assembly (0x459 + ret => 0x470 + ret); but shrank the async version (MoveNext 0x710 + ret => 0x659 + ret)

Interestingly, the smallest sync version came from caching the span to a local.  Caching span+length or length-only both resulted in even more assembly.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 4]
```diff
@@ -308,6 +310,27 @@ private static ArraySegment<byte> ReadToEnd(Stream stream)
...
+                // Read up to 3 bytes to see if it's the UTF-8 BOM
+                do
+                {
+                    // No need for checking for growth, the minimal rent sizes both guarantee it'll fit.
+                    Debug.Assert(rented.Length >= utf8Bom.Length);
+
+                    lastRead = stream.Read(
+                        rented,
+                        written,
+                        utf8Bom.Length - written);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: Why not read up to `rented.Length - written` to avoid an extra "small" read?

And then, we can just check if `rented.StartsWith(utf8Bom)`

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 5]
```diff
@@ -308,6 +310,27 @@ private static ArraySegment<byte> ReadToEnd(Stream stream)
...
+                // If we have 3 bytes, and they're the BOM, reset the write position to 0.
+                if (written == utf8Bom.Length &&
+                    utf8Bom.SequenceEqual(rented.AsSpan(0, utf8Bom.Length)))
+                {
+                    written = 0;
+                }
+
                 do
                 {
                     if (rented.Length == written)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: Just to confirm, this will never be true on the first iteration of the loop (if the stream contained a bom and we reset `written = 0`), correct?

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 6]
```diff
@@ -353,11 +376,14 @@ private static async
 
             try
             {
+                // Save the length to a local to be reused across awaits.
+                int utf8BomLength = JsonConstants.Utf8Bom.Length;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: Why is this necessary? Is it costly to access the span/length each time?

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 7]
```diff
@@ -367,6 +393,28 @@ private static async
...
+                // Read up to 3 bytes to see if it's the UTF-8 BOM
+                do
+                {
+                    // No need for checking for growth, the minimal rent sizes both guarantee it'll fit.
+                    Debug.Assert(rented.Length >= JsonConstants.Utf8Bom.Length);
+
+                    lastRead = await stream.ReadAsync(
+                        rented,
+                        written,
+                        utf8BomLength - written,
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: Again, to confirm, this will never be negative, correct? Maybe `Debug.Assert`

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs [hunk 8]
```diff
@@ -292,13 +292,15 @@ private static ArraySegment<byte> ReadToEnd(Stream stream)
             int written = 0;
             byte[] rented = null;
 
+            ReadOnlySpan<byte> utf8Bom = JsonConstants.Utf8Bom;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: Do you think `ReadToEnd` should be public available (maybe a static on stream/extension method)?

It is being used here as well (something we need to consider updating): https://github.com/dotnet/core-setup/blob/45f9401bf62faf0d3446cfd8681d35cc3487367a/src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.Utf8JsonReader.cs#L39-L89

File: src/System.Text.Json/tests/JsonDocumentTests.cs [hunk 9]
```diff
@@ -19,12 +19,27 @@ namespace System.Text.Json.Tests
...
     {
+        private static readonly byte[] Utf8Bom = { 0xEF, 0xBB, 0xBF };
+
         private static readonly Dictionary<TestCaseType, string> s_expectedConcat =
             new Dictionary<TestCaseType, string>();
 
         private static readonly Dictionary<TestCaseType, string> s_compactJson =
             new Dictionary<TestCaseType, string>();
 
+        public static IEnumerable<object[]> BadBOMCases { get; } =
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/JsonDocumentTests.cs]: Consider adding a test where the payload only contains the bom.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 4 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: The original version that I wrote just checked at the end if there was a BOM and changed the ReadOnlyMemory to use a non-zero start index; the problem is that the "I'm renting data on behalf of a document, clear it" doesn't know what the start index was.

The solutions that I envisioned:

* Make JsonDocument bigger by storing the start position of a temporary rent
* Make temporary rents clear the whole array before returning it, instead of just the written to portion
* Use Buffer.BlockCopy to move the data backwards back to position 0.
* This one

"small read" seemed the lowest impact.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 5 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: `rented` has a minimum size of 4 (and then, further, whatever the arraypool rounds up to from there).

If we don't reset, then it's (4+) vs 3, so not equal.  If we do reset then it's (4+) vs 0, so also not equal.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 6 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: The alternative to this is to make each place that uses it read the property and access the length field.  While the JIT could do magic there, sharplab said that accessing JsonConstants.Utf8Bom.Length each time made the async state machine's MoveNext be larger.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 7 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: I'm not sure what that assert would be guarding against... if a Stream implementation returns a number from ReadAsync bigger than count then pretty much everything ever touching a Stream is broken.

The guard against that would be ReadAsync itself, which should be guarding against a negative count.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 8 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: Maybe?  But if it was public it would not have this change; this is particular to us reading a stream (which is "bytes" and could be from File.OpenRead) and interpreting the data as UTF-8 text without going through the Encoding classes.

A more general fix (if we wanted to support UTF-16(BOM)) would be to ask StreamReader to do encoding sniffing, then translate all of the chars back through Encoding.UTF8.

DependencyModel would need the change if they similarly want to support BOM-prepended UTF-8 files; but if they don't write BOMs they wouldn't need it.

Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: > DependencyModel would need the change if they similarly want to support BOM-prepended UTF-8 files; but if they don't write BOMs they wouldn't need it.

 @eerhardt, does DependencyModel need to support files with the BOM present when reading?

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 6 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: Really? I would have thought the other way around since the local now needs to survive across the async/await boundary. cc @stephentoub

Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: [sharplab.io](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKEwCZtzMBmfAifSxgdlwG9zdf8HCzTAA5GAVgB0mZgFlEASwB2ACgCUXHny2YAnLhkBTOAAsA9gBMCyxQYDuuYAE84BgNoFs2ALoxcN+4QSAJIA8hKGqKYATo4AynBRBoioaqqkZFoAvpq8ObgADlHyAG6ILoxCooSS0gA8SnAAfPpGZpbKTi6uXriJii7mvoGhEvGJybjICUmoqnncGVp8DbgArnAAZsIAQqaoADIGigDmJrgAvLh06UvL/bi2RXAuihe42De3uCsANihwACUkuY3h88lpzKZwXwFl9bn8pkDECDLrpJtNkhIkeYAILIRyKADGymhcN4fQGMFJZMe8meRypizJS3WW12ByOp2MuAAtA8ni9VBIAMKmRQbeTHVaJHG2BRwZQbRA/ZAGNLkalLWn014AakuCMBwM+t0yD2M8h+BlwykN2NwzTwADInfy6S9cLU1psdntDicTOqmUtqZgOHbjXlssG8oUSmVrYJ8FUCDUIPV+s1DCYLJQOs43D0KQZBoxgmExjN0eNZvNqSttR7LmDg1pfv97c2TUtIdTYczeBHkW80VMa1jgXiCcTNV9i4NZ7dGwzF0sGHzl4ohaLxZLpQZZfLFcrVUHV26dbh9bgh+Zu1lzZbrbaO8CHe9cC6Lx6vXQg3DQ3DV9kW7aNMiAA==)

Before - 32 bytes:
```text
Program+<Method1>d__1.SetStateMachine(System.Runtime.CompilerServices.IAsyncStateMachine)
    L0000: sub rsp, 0x28
    L0004: add rcx, 0x20
```
After - 24 bytes:
```text
Program+<Method2>d__2.SetStateMachine(System.Runtime.CompilerServices.IAsyncStateMachine)
    L0000: sub rsp, 0x28
    L0004: add rcx, 0x18
...
```
+ 4 fewer instructions in MoveNext (if the length was just a regular integer literal).
Accessing the span property + length does increase the instruction count.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs (hunk 8 above)
Comment (eerhardt) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: It just used `JsonTextReader` from Newtonsoft previously. Does that handle BOMs?

Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs]: > It just used JsonTextReader from Newtonsoft previously. Does that handle BOMs?

Technically, no.  But the StreamReader adapter between the Stream and JsonTextReader does :smile:.  So this change _should_ be ported.
